### PR TITLE
Kotlin implementation

### PR DESCRIPTION
**Pull request by arctouch-paularosa**
Migrate Java code to kotlin

* Add extensions functions to code
* Uses Parcelize to build parcelable classes
* Refactor code to avoid requesting data without a previous response

### CODE REVIEW

**Comment by arctouch-ianribas** [comment]
Fix all compiler warnings ...

File: app/src/main/java/net/paulacr/movies/feature/movieslist/adapter/ListMoviesAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+            val movie = movies[position]
+            val viewHolder = holder as ViewHolderList
+
+            viewHolder.bind(movie)
+            holder.itemView.setOnClickListener { onListItemClick.onClick(movie) }
+        }
+    }
+
+    override
+    fun getItemCount(): Int {
```
Comment (arctouch-ianribas) [inline app/src/main/java/net/paulacr/movies/feature/movieslist/adapter/ListMoviesAdapter.kt]: This function (and the next) would look good as function expressions:

```kotlin
override fun getItemCount(): Int = if (movies.isEmpty()) 1 else movies.size
```

File: app/src/main/java/net/paulacr/movies/feature/movieslist/MovieListPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+                BiFunction { genres: Map<String, String>, movies: Results ->
+                    combineMoviesWithGenres(genres, movies.results)
+                })
+
+                .observeOn(AndroidSchedulers.mainThread())
+                .subscribe(
+                        {
+                            listMoviesView.showMovies(it)
+                            canInvalidatePagesCount = true
```
Comment (arctouch-ianribas) [inline app/src/main/java/net/paulacr/movies/feature/movieslist/MovieListPresenter.kt]: This variable seems to never be used.

File: app/src/main/java/net/paulacr/movies/feature/movieslist/MovieListPresenter.kt [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        }
+        return resultMovies
+
+    }
+
+    private fun getGenres(genresMovieId: MutableList<String>, map: Map<String, String>): MutableList<String> {
+        var genresNamesList = mutableListOf<String>()
+
+        for (id: String in genresMovieId) {
+            val name = map[id]
```
Comment (arctouch-ianribas) [inline app/src/main/java/net/paulacr/movies/feature/movieslist/MovieListPresenter.kt]: Variable `name` is not used.

File: app/src/main/java/net/paulacr/movies/feature/moviesdetail/MovieDetailActivity.kt [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+ * Created by paularosa on 07/03/18.
+ */
+
+class MovieDetailActivity: AppCompatActivity(), MoviesDetailContract.View {
+
+    companion object {
+        const val KEY_MOVIE = "key_movie"
+
+        fun newIntent(context: Context, movie: Movie): Intent {
+            val intent = Intent(context, MovieDetailActivity::class.java)
```
[the remark this answers is not in the document]
Comment (paulacr) [reply app/src/main/java/net/paulacr/movies/feature/moviesdetail/MovieDetailActivity.kt]: Excellent

**Pull request merged**